### PR TITLE
SourceMap#fromSource should preserve columns

### LINKED ISSUE #5: SourceMap.forSource removes column information.
I'm pretty sure you just want a mapping from position 1 to position 1, rather than iterating through all the lines as you are.

This causes plumber to only report line numbers and not column numbers.

### PR DESCRIPTION
**Pull request by insidewhy**
This fixes plumber losing column information during certain operations.

Fixes https://github.com/plumberjs/mercator/issues/5

### CODE REVIEW

**Comment by insidewhy** [comment]
You guys should totally pull this, losing column information is really bad!

File: test/test.js [hunk 1]
```diff
@@ -71,17 +71,8 @@ describe('SourceMap', function() {
...
-            for (var line = 1; line <= numLines; line++) {
-                consumer.originalPositionFor({line: line, column: 0}).should.deep.equal({
-                    source: path,
-                    line: line,
-                    column: 0,
-                    name: null
-                });
-            }
+            // Should map according to javascript tokens.
+            map.mappings.should.equal('AAAA,IAAI,EAAE,EAAE,CAAC;AACT,IAAI,EAAE,EAAE,CAAC;;;;AAIT,SAAS,GAAG,CAAC,CAAC,EAAE;IACZ,OAAO,EAAE,EAAE,CAAC;AAChB;;AAEA,IAAI,EAAE,EAAE,GAAG,CAAC,CAAC,CAAC');
```
Comment (theefer) [inline test/test.js:75]: It'd be more helpful to have a test like above that makes it clearer what part of the test may fail.

**Comment by theefer** [comment]
Could you add some tests that show what information is lost, and how this PR fixes the issue?

File: test/test.js (hunk 1 above)
Comment (insidewhy) [inline test/test.js:75]: Yeah, the test would have to scan the file with esprima itself and match on a token by token basis, maybe re-using bits of the source map API to work out what the data is. Essentially you'd be repeating the tokenisation process into the test, which would be a bit silly. So I think it's better like this. Kinda.

**Comment by insidewhy** [comment]
The test I amended demonstrates what is lost by the way it was amended to reflect all tokens rather than only preserving lines.

**Comment by theefer** [comment]
It's not really clear what these mappings are and where you got them from though. Could it be more explicit?

And can you give an example of a setup where having all these initial mappings is necessary to have the correct maps throughout?

**Comment by insidewhy** [comment]
The issue is that your way of creating an identity mapping is wrong. Source maps work on tokenised data, mercator is mapping from column 0 to column 0 for each line and losing granularity in the process. The create an identity source mapping the file has to be tokenised, it's kinda a deficiency of the v3 implementation of source maps... that even an identity mapping requires like 4 bytes per token.

It's pretty obvious really after you understand... I guess you've noticed that your plumber pipelines lose column information when the first step comes from say... glob (a non-source-map producing operation)?

That data is the identity mapping for the test source file you provide. To reconstruct it inside the test makes the test as large as the implementation, it's not a good idea. Since it's so trivially obvious it can be taken for granted that the source map content is correct.

**Comment by insidewhy** [comment]
I guess it might be cool if the test demonstrated the process of creating an identity source map... but to be honest a ~~content~~ comment linking to the relevant lines of code would do the same job since they'd be identical.

**Comment by insidewhy** [comment]
Really disappointed with this one, we've had to rip out mercator from everything now and rewrite a working version. GIven we've abandoned plumber/mercator and you don't seem to understand this issue may as well close it.

**Comment by insidewhy** [comment]
If anybody actually uses plumber/mercator (and pity on those that waste as much time with it as we have) it'll just get re-opened again anyway because it's so fundamental.

**Comment by theefer** [comment]
Sorry for the lack of activity, I've been busy with other things and haven't been able to commit to Plumber enough.

I do understand the issue. What I was asking was to rewrite the tests in such a way that they are human-readable, or add an extra human-readable test that demonstrate the mapping of the tokens. Otherwise, we're left comparing completely opaque mapping encoded strings, which will make it almost impossible to understand if anything goes slightly wrong (e.g. off by one, etc).

**Pull request closed without merging**